# BiocNYC-ChIPSeqSpike

# Performing ChIP-Seq data scaling with spike-in control in R/Bioconductor

04/19/2018

## Introduction

In this workshop, we will look at performing ChIP-Seq data scaling with spike-in controls. We will use 'ChIPSeqSpike' that is a new package at an early stage of development. You will learn how to perform the spike-in scaling properly and to visualize the different transformations in R. You will also be able to compare the effect of an inhibitor treatment on the Histone 3 Lysine 79 di-methylation (H3K79me2).

This workshop will not cover the data pre-processing and already processed BAM and BIGWIG files will be provided.


### Covering today

* Background on ChIP-Seq spike-in (see [pdf]( BiocNYC-ChIPSeqSpike/BioconductorMeetupAp2018.pdf ))

* Construction of objects containing all files necessary for proper scaling

* Computation and evaluation of scaling factors 

* 
